crates/ruff_python_parser/src/parser/mod.rs

```diff
@@ -32,6 +41,12 @@ pub(crate) struct Parser<'src> {
     /// Stores all the syntax errors found during the parsing.
     errors: Vec<ParseError>,
 
+    /// Stores non-fatal syntax errors found during parsing, such as version-related errors and
+    /// errors detected by the Python compiler.
+    syntax_errors: Vec<SyntaxError>,
+
+    syntax_error_state: SyntaxErrorState,
```

Comment:
My suggestion here is to only handle version specific syntax errors in the parser. I'd still recommend handling semantic (compiler errors) in a separate pass outside the parser.